File: src/main/java/com/cupoftea/discordmc/commands/DiscordMCCommands.java
package com.cupoftea.discordmc.commands;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

import org.bukkit.ChatColor;
import org.bukkit.command.CommandSender;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.entity.Player;

import com.cupoftea.discordmc.DiscordMCPlugin;

import co.aikar.commands.BaseCommand;
import co.aikar.commands.annotation.CommandAlias;
import co.aikar.commands.annotation.CommandPermission;
import co.aikar.commands.annotation.Default;
import co.aikar.commands.annotation.Description;
import co.aikar.commands.annotation.Subcommand;

@CommandAlias("discordmc|dmc")
public class DiscordMCCommands extends BaseCommand {

    private final DiscordMCPlugin plugin;
    private final Map<String, PendingLink> pendingLinks = new HashMap<>();

    public DiscordMCCommands(DiscordMCPlugin plugin) {
        this.plugin = plugin;
    }

    @Default
    @Description("Shows DiscordMC commands")
    public void showHelp(CommandSender sender) {
        sender.sendMessage(ChatColor.GOLD + "=== DiscordMC Commands ===");
        sender.sendMessage(ChatColor.YELLOW + "/discordmc" + ChatColor.WHITE + " - Show this help message");
        sender.sendMessage(ChatColor.YELLOW + "/discordmc reload" + ChatColor.WHITE + " - Reload the plugin configuration");
        sender.sendMessage(ChatColor.YELLOW + "/discordmc status" + ChatColor.WHITE + " - Check the Discord connection status");
        sender.sendMessage(ChatColor.YELLOW + "/discordmc togglelinking" + ChatColor.WHITE + " - Toggle the ability to link accounts");
        sender.sendMessage(ChatColor.YELLOW + "/link <code>" + ChatColor.WHITE + " - Link your Minecraft account to Discord");
        sender.sendMessage(ChatColor.YELLOW + "/discord" + ChatColor.WHITE + " - Show your linked Discord account");
        sender.sendMessage(ChatColor.YELLOW + "/unlink" + ChatColor.WHITE + " - Unlink your Minecraft account from Discord");
    }

    @Subcommand("reload")
    @Description("Reloads the DiscordMC configuration")
    @CommandPermission("discordmc.reload")
    public void reloadConfig(CommandSender sender) {
        sender.sendMessage(ChatColor.YELLOW + "Reloading DiscordMC configuration...");
        if (plugin.reload()) {
            sender.sendMessage(ChatColor.GREEN + "DiscordMC configuration reloaded successfully!");
        } else {
            sender.sendMessage(ChatColor.RED + "Failed to reload DiscordMC configuration. Check console for details.");
        }
    }

    @Subcommand("status")
    @Description("Checks the status of the Discord connection")
    @CommandPermission("discordmc.status")
    public void checkStatus(CommandSender sender) {
        boolean isConnected = plugin.getDiscordManager().isConnected();
        sender.sendMessage(ChatColor.YELLOW + "Discord connection status: " + 
                           (isConnected ? ChatColor.GREEN + "Connected" : ChatColor.RED + "Disconnected"));
    }

    @Subcommand("togglelinking")
    @CommandPermission("discordmc.togglelinking")
    @Description("Toggle the ability to link accounts")
    public void toggleLinking(CommandSender sender) {
        FileConfiguration config = plugin.getConfig();
        boolean currentState = config.getBoolean("allow-linking", true);
        config.set("allow-linking", !currentState);
        plugin.saveConfig();

        String state = !currentState ? "enabled" : "disabled";
        sender.sendMessage(ChatColor.GREEN + "Account linking has been " + state + ".");
    }

    @CommandAlias("link")
    @Description("Link your Minecraft account to Discord")
    public void linkAccount(Player player, String code) {
        if (!plugin.getConfig().getBoolean("allow-linking", true)) {
            player.sendMessage(ChatColor.RED + "Account linking is currently disabled.");
            return;
        }

        PendingLink pendingLink = pendingLinks.remove(code);
        if (pendingLink != null) {
            if (!player.getName().equalsIgnoreCase(pendingLink.minecraftUsername)) {
                player.sendMessage(ChatColor.RED + "This link code is for a different Minecraft username.");
                return;
            }

            String discordId = pendingLink.discordId;
            String existingLink = plugin.getMongoDBManager().getDiscordId(player.getUniqueId());
            if (existingLink != null) {
                player.sendMessage(ChatColor.RED + plugin.getConfigManager().getMessage("minecraft.link-already-linked"));
                return;
            }
            
            UUID existingMinecraftUUID = plugin.getMongoDBManager().getMinecraftUUID(discordId);
            if (existingMinecraftUUID != null) {
                player.sendMessage(ChatColor.RED + plugin.getConfigManager().getMessage("minecraft.link-discord-already-linked"));
                return;
            }
            
            plugin.getMongoDBManager().linkAccount(player.getUniqueId(), player.getName(), discordId);
            player.sendMessage(ChatColor.GREEN + plugin.getConfigManager().getMessage("minecraft.link-success"));
            plugin.getDiscordManager().getDiscordListener().assignLinkedRole(discordId);

            String successMessage = plugin.getConfigManager().getMessage("discord.dm-link-success").replace("{username}", player.getName());
            plugin.getDiscordManager().sendDirectMessage(discordId, successMessage);
        } else {
            player.sendMessage(ChatColor.RED + plugin.getConfigManager().getMessage("minecraft.link-invalid-code"));
        }
    }

    @CommandAlias("discord")
    @Description("Show your linked Discord account")
    public void showLinkedAccount(Player player) {
        String discordId = plugin.getMongoDBManager().getDiscordId(player.getUniqueId());
        if (discordId != null) {
            String discordUsername = plugin.getDiscordManager().getDiscordUsername(discordId);
            if (discordUsername != null) {
                player.sendMessage(ChatColor.GREEN + "Your Minecraft account is linked to Discord user: " + discordUsername);
            } else {
                player.sendMessage(ChatColor.RED + "Failed to retrieve Discord username. Please try again later.");
            }
        } else {
            player.sendMessage(ChatColor.RED + "Your Minecraft account is not linked to a Discord account.");
        }
    }

    @CommandAlias("unlink")
    @Description("Unlink your Minecraft account from Discord")
    public void unlinkAccount(Player player) {
        String discordId = plugin.getMongoDBManager().unlinkAccount(player.getUniqueId());
        if (discordId != null) {
            player.sendMessage(ChatColor.GREEN + "Your Minecraft account has been unlinked from Discord.");
            plugin.getDiscordManager().getDiscordListener().removeLinkedRole(discordId);
            
            String unlinkMessage = "Your Discord account has been unlinked from the Minecraft account: **" + player.getName() + "**";
            plugin.getDiscordManager().sendDirectMessage(discordId, unlinkMessage);
        } else {
            player.sendMessage(ChatColor.RED + "Your Minecraft account is not linked to a Discord account.");
        }
    }

    public String generateLinkCode(String discordId, String minecraftUsername) {
        String code = UUID.randomUUID().toString().substring(0, 6);
        pendingLinks.put(code, new PendingLink(discordId, minecraftUsername));
        return code;
    }

    private static class PendingLink {
        final String discordId;
        final String minecraftUsername;

        PendingLink(String discordId, String minecraftUsername) {
            this.discordId = discordId;
            this.minecraftUsername = minecraftUsername;
        }
    }
}

File: src/main/kotlin/com/cupoftea/discordmc/commands/DiscordMCCommands.kt
package com.cupoftea.discordmc.commands

import cloud.commandframework.annotations.CommandMethod
import cloud.commandframework.annotations.CommandPermission
import cloud.commandframework.annotations.CommandDescription
import cloud.commandframework.annotations.Argument
import com.cupoftea.discordmc.DiscordMCPlugin
import com.cupoftea.discordmc.minecraft.service.AccountLinkService
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import com.cupoftea.discordmc.utils.MessageUtils.toComponent

class DiscordMCCommands(plugin: DiscordMCPlugin) {
    private val configurationService = DiscordMCPlugin.configurationService
    private val accountLinkService = DiscordMCPlugin.accountLinkService
    private val discordService = DiscordMCPlugin.discordService

    @CommandMethod("discordmc|dmc")
    @CommandDescription("Shows DiscordMC commands")
    fun showHelp(sender: CommandSender) {
        sender.sendMessage("""
            <gold>=== DiscordMC Commands ===</gold>
            <yellow>/discordmc</yellow> <white>- Show this help message</white>
            <yellow>/discordmc reload</yellow> <white>- Reload the plugin configuration</white>
            <yellow>/discordmc status</yellow> <white>- Check the Discord connection status</white>
            <yellow>/discordmc togglelinking</yellow> <white>- Toggle the ability to link accounts</white>
            <yellow>/link <code></yellow> <white>- Link your Minecraft account to Discord</white>
            <yellow>/discord</yellow> <white>- Show your linked Discord account</white>
            <yellow>/unlink</yellow> <white>- Unlink your Minecraft account from Discord</white>
        """.trimIndent().toComponent())
    }

    @CommandMethod("discordmc reload")
    @CommandPermission("discordmc.reload")
    @CommandDescription("Reloads the DiscordMC configuration")
    fun reloadConfig(sender: CommandSender) {
        sender.sendMessage("<yellow>Reloading DiscordMC configuration...</yellow>".toComponent())
        configurationService.reload()
        sender.sendMessage("<green>DiscordMC configuration reloaded successfully!</green>".toComponent())
    }

    @CommandMethod("discordmc status")
    @CommandPermission("discordmc.status")
    @CommandDescription("Checks the status of the Discord connection")
    fun checkStatus(sender: CommandSender) {
        val isConnected = discordService.isInitialized()
        sender.sendMessage("§eDiscord connection status: ${if (isConnected) "§aConnected" else "§cDisconnected"}")
    }

    @CommandMethod("discordmc togglelinking")
    @CommandPermission("discordmc.togglelinking")
    @CommandDescription("Toggle the ability to link accounts")
    fun toggleLinking(sender: CommandSender) {
        val currentState = configurationService.getBoolean("allow-linking")
        configurationService.set("allow-linking", !currentState)
        val state = if (!currentState) "enabled" else "disabled"
        sender.sendMessage("§aAccount linking has been $state.")
    }

    @CommandMethod("link <code>")
    @CommandDescription("Link your Minecraft account to Discord")
    suspend fun linkAccount(player: Player, @Argument("code") code: String) {
        if (!configurationService.getBoolean("allow-linking")) {
            player.sendMessage("<red>Account linking is currently disabled.</red>".toComponent())
            return
        }

        val pendingLink = accountLinkService.getPendingLink(code) ?: run {
            player.sendMessage("§c${configurationService.getMessage("minecraft.link-invalid-code")}")
            return
        }

        if (player.name != pendingLink.minecraftUsername) {
            player.sendMessage("§cThis link code is for a different Minecraft username.")
            return
        }

        when (val result = accountLinkService.linkAccount(
            player.uniqueId,
            player.name,
            pendingLink.discordId
        )) {
            is AccountLinkService.LinkResult.Success -> {
                player.sendMessage("§a${configurationService.getMessage("minecraft.link-success")}")
                configurationService.getString("discord.linked-role-id")?.let { roleId ->
                    discordService.assignRole(pendingLink.discordId, roleId)
                }
                discordService.sendPrivateMessage(
                    pendingLink.discordId,
                    configurationService.getMessage("discord.dm-link-success", "username" to player.name).toString()
                )
            }
            is AccountLinkService.LinkResult.AlreadyLinkedDiscord -> {
                player.sendMessage("§c${configurationService.getMessage("minecraft.link-discord-already-linked")}")
            }
            is AccountLinkService.LinkResult.AlreadyLinkedMinecraft -> {
                player.sendMessage("§c${configurationService.getMessage("minecraft.link-already-linked")}")
            }
        }
    }

    @CommandMethod("discord")
    @CommandDescription("Show your linked Discord account")
    suspend fun showLinkedAccount(player: Player) {
        val discordId = accountLinkService.getDiscordId(player.uniqueId)
        if (discordId != null) {
            val discordUsername = discordService.getUsernameById(discordId)
            if (discordUsername != null) {
                player.sendMessage("§aYour Minecraft account is linked to Discord user: $discordUsername")
            } else {
                player.sendMessage("§cFailed to retrieve Discord username. Please try again later.")
            }
        } else {
            player.sendMessage("§cYour Minecraft account is not linked to a Discord account.")
        }
    }

    @CommandMethod("unlink")
    @CommandDescription("Unlink your Minecraft account from Discord")
    suspend fun unlinkAccount(player: Player) {
        when (val result = accountLinkService.unlinkAccount(player.uniqueId)) {
            is AccountLinkService.UnlinkResult.Success -> {
                player.sendMessage("§a${configurationService.getMessage("minecraft.unlink-success")}")
                configurationService.getString("discord.linked-role-id")?.let { roleId ->
                    discordService.removeRole(result.discordId, roleId)
                }
                discordService.sendPrivateMessage(
                    result.discordId,
                    configurationService.getMessage("discord.dm-unlink-success", "username" to player.name).toString()
                )
            }
            is AccountLinkService.UnlinkResult.NotLinked -> {
                player.sendMessage("§c${configurationService.getMessage("minecraft.unlink-not-linked")}")
            }
        }
    }
} 